Fix the multicore performance of mapping from an instruction pointer to a MethodDesc

davidwrighton opened the pull request:
Mapping from an instruction pointer to a MethodDesc is a critical component of the runtime used in many places, notably diagnostics, slow path delegate creation, and stackwalking.

This change provides a dramatic improvement in the performance of that logic through several techniques.

1. The mapping from IP to range of interesting memory is done via a tree structure resembling that of a page table. The result of this is that in addition to reducing the locking logic, the cost of looking up in the presence of many different loaded assemblies will be reduced to a nearly constant time.
2. That tree structure is configured so that when accessing memory which will never be freed in the lifetime of the process, the memory can be accessed without taking any locks whatsoever.
3. The code map is enhanced to include not only the code generated by the JIT/R2R code, but also to include the fixup and stub precode stubs.
4. In addition, performance improvement was done to improve the performance of slow path delegate creation in particular, by reordering which checks are done, and by writing a simplified signature parse routine for computing the number of arguments to a function.

Performance of this was tested in both the EH stackwalking scenario as well as the delegate slow path creation scenario, but the delegate logic is what will be most visible when this is checked in. (See PR #79471 for details of the additional changes necessary to take advantage of this work when doing EH) (#8393 describes the potential product impact of just improving the delegate slow path creation code)

For the delegate creation scenario the perf impact was measured to be quite substantial. These numbers are in ms to create a constant number of delegates on each thread used. Smaller is better. The test was run on a 6 core machine.

| Threads | Without fix | With this PR
| ------------- | ------------- | ----- |
| 1  | 840 |  313|
| 2 | 1977 | 316 |
| 3 | 3290 | 325 |
| 4| 4259 | 344 |
| 5 | 5140 | 351 |
| 6 | 5872 | 374|
| 7 | 6463 | 442|
| 8 | 7046 | 499|
| 9 | 7648 | 547|
| 10 | 8241 | 627|
| 11 | 8851 | 749|
| 12 | 9595 | 773|

Fixes #8393
Replaces the controversial logic in PR #68607

davidwrighton committed:
Experimental

davidwrighton committed:
Experiment

davidwrighton committed:
More progress

davidwrighton committed:
Increase accuracy of comments around range section locking
- This is improving docs on the old scheme
- Also remove the ngen specific logic

davidwrighton committed:
Changes to RangeListMap to make it work
- still with the wrong locking though...

davidwrighton committed:
Add low bit check indirection

davidwrighton committed:
Rework to passing down the current locking status

davidwrighton committed:
Allocation of RangeSection and freeing added, as well as testing for the collectible handling characteristics

davidwrighton committed:
Fix some naming

davidwrighton committed:
Miraculously it builds!

davidwrighton committed:
May actually work. DAC access known to be broken.

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into rangelistmap

EgorBo commented:
Nice! Does it make `ExecutionEngine::IsReadyToRun(PCODE)` cheap?

davidwrighton commented:
@EgorBo, yes, I do expect that this will make IsReadyToRun very cheap. (Probably 5-6 memory loads from separate pages plus some math, so repeated calls should be extremely cheap, and random calls will be somewhat expensive, but not bad.) The implementation as written has a risk of entering a lock under some circumstances, but after sleeping on it, I think we could get rid of that for the IsReadyToRun logic, and be assured that IsReadyToRun is cheap. What do you think you can optimize?

EgorBo commented:
> What do you think you can optimize?

I'd like to sort of de-prioritize R2R'd code for tier1 promotion in favor of non R2R'd one.

davidwrighton committed:
Fix handling of unwind table

davidwrighton committed:
Seems to work well on WinX64

Now that it is profileable, there are tweaks to be made

davidwrighton commented:
@EgorBo, fyi, `ExecutionManager::IsReadyToRun` is slower than `ExecutionManager::FindReadyToRunModule`. Also, I've found there may be a slight bug in the current implementation of the locking strategy for those 2 methods, so I'm addressing that as well.

davidwrighton committed:
Fix performance of the DelegateConstruct method
- Implement a fast path for determining that an instruction pointer is a FixupPrecode/StubPrecode. It turns out those are more likely, than actual jitted code (as we typically have wrapper stubs of some form at this point).
- Implement a fast path for computing the number of arguments to a method instead of using MetaSig.
- Remove fancy locking scheme from RangeSectionMap. The existing fancy locking scheme works fine.
- Fix bug in ExecutionManager::IsReadyToRun and ExecutionManager::FindReadyToRunModule where locks were not used

davidwrighton committed:
It should build everywhere now

davidwrighton committed:
Delete temporary bits

davidwrighton committed:
Fix Windows builds

davidwrighton committed:
Fix contract issues

davidwrighton committed:
Fix CodeRangeMapRangeList to handle interval correctly

davidwrighton committed:
It builds on Linux!

davidwrighton committed:
Make it even more lock free

davidwrighton committed:
Remove confusion around Range interval specification Some uses were relying on [,) semantics and some were [,]

davidwrighton committed:
Attempt to appease the GCC build

davidwrighton committed:
Fix the math around fragment handling and index calculations

davidwrighton committed:
Make the RangeSectionMap available to the DAC

davidwrighton committed:
Move to 5 levels for 64bit RangeSectionMap. 8KB chunks are a waste of memory

davidwrighton committed:
Remove unnecessary memory dereferences
- Use the byte array trick to remove unneeded dereferences.

davidwrighton committed:
Tweak algorithm to allow for R2R checks to be entirely lock-free

davidwrighton committed:
Update comment

davidwrighton committed:
Enable support for freeing portion of the CodeRangeMap which are no longer used after code is collected

davidwrighton committed:
Fix GCC noticed signed/unsigned mismatch

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+
+        TADDR FragmentToPtr(RangeSectionFragment* fragment)
+        {
+            TADDR ptr = (TADDR)fragment;
```
```suggestion
            TADDR ptr = dac_cast<TADDR>(fragment);
```
or ifdef out the method for DACCESS_COMPILE

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+    private:
+        TADDR _ptr;
+
+        TADDR FragmentToPtr(RangeSectionFragment* fragment)
```
```suggestion
        static TADDR FragmentToPtr(RangeSectionFragment* fragment)
```

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+    private:
+        TADDR _ptr;
+
+        TADDR LevelToPtr(TPtr level, bool collectible)
```
```suggestion
        static TADDR LevelToPtr(TPtr level, bool collectible)
```

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+
+        TADDR LevelToPtr(TPtr level, bool collectible)
+        {
+            TADDR ptr = (TADDR)level;
```
```suggestion
            TADDR ptr = dac_cast<TADDR>(level);
```

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+    };
+
+    // Unlike a RangeSection, a RangeSectionFragment cannot span multiple elements of the last level of the RangeSectionMap
+    // Always allocated via memset/free
```
```suggestion
    // Always allocated via malloc/free
```

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
     };
 
-    DWORD               flags;
+#ifdef TARGET_64BIT
+    static const uintptr_t entriesPerMapLevel = 256;
+#else
+    static const uintptr_t entriesPerMapLevel = 256;
+#endif
```
```suggestion
    static const uintptr_t entriesPerMapLevel = 256;
```

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+        return buf;
+    }
+
+    uintptr_t EffectiveBitsForLevel(TADDR address, uintptr_t level)
```
```suggestion
    static uintptr_t EffectiveBitsForLevel(TADDR address, uintptr_t level)
```

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+    uintptr_t RangeSectionFragmentCount(PTR_RangeSection pRangeSection)
+    {
+        uintptr_t rangeSize = pRangeSection->_range.RangeSize();
+        if (rangeSize == 0)
```
How can we end up with RangeSection here?

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+
+struct RangeSectionMapData
+{
+    BYTE Data[sizeof(RangeSectionMap)];
```
```suggestion
    TADDR Data[sizeof(RangeSectionMap) / sizeof(TADDR)];
```
We need to make sure that the compiler places the static at sufficiently aligned address.

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+        return fragmentCount;
+    }
+
+    TADDR IncrementAddressByMaxSizeOfFragment(TADDR input)
```
```suggestion
    static TADDR IncrementAddressByMaxSizeOfFragment(TADDR input)
```

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+        memset(&_topLevelData, 0, sizeof(_topLevelData));
+    }
+
+    bool Init()
+    {
+        return true;
+    }
+
```
```suggestion
```
Not needed?

jkotas commented on src/coreclr/vm/codeman.h:
```diff
@@ -608,44 +610,926 @@ class UnwindInfoTable {
+        uintptr_t topLevelIndex = EffectiveBitsForLevel(address, --level);
+        auto nextLevelAddress = &(GetTopLevel()[topLevelIndex]);
+#ifdef TARGET_64BIT
+        auto _RangeSectionL4 = nextLevelAddress;
```
Nit: Do not use `_` for locals

jkotas reviewed (approved):
LGTM modulo feedback. Nice work!

davidwrighton committed:
Fix next GCC noticed signed/unsigned mismatch

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into rangelistmap

davidwrighton committed:
Adjust to feedback

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into rangelistmap
